src/backend/distributed/connection/locally_reserved_shared_connection_stats.c

```diff
@@ -0,0 +1,487 @@
+	key.port = nodePort;
+	key.databaseOid = databaseOid;
+
+	bool found = false;
+	ReservedConnectionCounterHashEntry *entry =
+		(ReservedConnectionCounterHashEntry *) hash_search(
+			SessionLocalReservedConnectionCounters, &key, HASH_FIND, &found);
+
+	/* we should have already allocated */
+	if (!found)
+	{
+		ereport(ERROR, (errmsg("BUG: untracked reserved connection"),
+						errhint("Set citus.max_shared_pool_size TO -1 to "
+								"disable reserved connection counters")));
+	}
+
+	entry->reservedConnectionCount++;
+
+	ereport(DEBUG5, (errmsg("Incremented shared reserved connection to node "
+							"%s:%d to %d", entry->key.hostname, entry->key.port,
+							entry->reservedConnectionCount)));
+}
+
+
+/*
+ * This function is intended to be called when the session has reserved connections
+ * but not used them, and we are in a good position to get rid of them.
+ */
+void
+DeallocateAllReservedConnections(void)
+{
+	HASH_SEQ_STATUS status;
+	ReservedConnectionCounterHashEntry *entry;
+
+	hash_seq_init(&status, SessionLocalReservedConnectionCounters);
+	while ((entry = (ReservedConnectionCounterHashEntry *) hash_seq_search(&status)) != 0)
+	{
+		while (entry->reservedConnectionCount > 0)
+		{
+			DecrementSharedConnectionCounter(entry->key.hostname, entry->key.port);
+			entry->reservedConnectionCount--;
+
+			ereport(DEBUG5, (errmsg("Decrement shared reserved connection to node "
+									"%s:%d to %d", entry->key.hostname, entry->key.port,
+									entry->reservedConnectionCount)));
+		}
+
+		bool found = false;
+
+		/*
+		 * We cleaned up all the entries because we may not need reserved connections
+		 * in the next iteration.
+		 */
+		hash_search(SessionLocalReservedConnectionCounters, entry, HASH_REMOVE, &found);
+		Assert(found);
+	}
+}
+
+
+/*
+ * DecrementReservedConnection decrements the reserved counter for the given host.
+ */
+void
+DecrementReservedConnection(const char *hostName, int nodePort, Oid databaseOid)
+{
+	ReservedConnectionCounterHashKey key;
+
+	strlcpy(key.hostname, hostName, MAX_NODE_LENGTH);
+	key.port = nodePort;
+	key.databaseOid = databaseOid;
+
+	bool found = false;
+	ReservedConnectionCounterHashEntry *entry =
+		(ReservedConnectionCounterHashEntry *) hash_search(
+			SessionLocalReservedConnectionCounters, &key, HASH_FIND, &found);
+
+	if (!found)
+	{
+		ereport(ERROR, (errmsg("BUG: untracked reserved connection"),
+						errhint("Set citus.max_shared_pool_size TO -1 to "
+								"disable reserved connection counters")));
+	}
+
+	entry->reservedConnectionCount--;
+
+	ereport(DEBUG5, (errmsg("Decrement shared reserved connection to node %s:%d to %d",
+							hostName, nodePort, entry->reservedConnectionCount)));
+}
+
+
+/*
+ * ReserveSharedConnectionCounterForAllPrimaryNodesIfNeeded is a wrapper around
+ * ReserveSharedConnectionCounterForNodeListIfNeeded.
+ */
+void
+ReserveSharedConnectionCounterForAllPrimaryNodesIfNeeded(void)
+{
+	List *primaryNodeList = ActivePrimaryNonCoordinatorNodeList(NoLock);
```

Comment:
the coordinator may have shards

it might also make sense to take a lock such that we don't add any new nodes during the copy